quarkchain/cluster/shard.py

```diff
@@ -705,28 +763,34 @@ def broadcast_tx_list(self, tx_list, source_peer=None):
                 Logger.error_exception()
                 return False, coinbase_amount_list
 
-            # block already existed in local shard state
-            # but might not have been propagated to other shards and master
-            # let's make sure all the shards and master got it before return
-            if xshard_list is None:
-                future = self.add_block_futures.get(block_hash, None)
-                if future:
-                    existing_add_block_futures.append(future)
-            else:
-                prev_root_height = self.state.db.get_root_block_header_by_hash(
-                    block.header.hash_prev_root_block
-                ).height
-                block_hash_to_x_shard_list[block_hash] = (xshard_list, prev_root_height)
-                self.add_block_futures[block_hash] = self.loop.create_future()
+            prev_root_height = self.state.db.get_root_block_header_by_hash(
+                block.header.hash_prev_root_block
+            ).height
+            block_hash_to_x_shard_list[block_hash] = (xshard_list, prev_root_height)
+            self.add_block_futures[block_hash] = self.loop.create_future()
+            block_header_list.append(block.header)
+            coinbase_amount_map_list.append(block.header.coinbase_amount_map)
 
         await self.slave.batch_broadcast_xshard_tx_list(
             block_hash_to_x_shard_list, block_list[0].header.branch
         )
+        check(len(coinbase_amount_map_list) == len(block_header_list))
+        await self.slave.send_minor_block_header_list_to_master(
```

Comment:
help me understand this part:

if the cluster shuts down here, those blocks remain uncommitted, so the logic will redo during recovery; in previous code, without #483, I can see xshard list broadcasting will be skipped thus causing issues, but after #483, shard will still broadcast xshard list, right? if yes what's the benefit of 2PC?

Reply:
After #483, the xshard list may not be broadcasted if the block is already added to the slave.  However, #483 will add the minor headers to validated pool in the master - assuming the slaves receive x-shard deposits, but actually, it is not.